### PR TITLE
impute_sex, now 100% python

### PR DESCRIPTION
**Pull request by danking**
cc: @tpoterba I added some new type check stuff

I added `round` `ceil` and `floor` as I was writing tests and then I didn't need them. They still seem useful, so I included them.

I had to add a tolerance parameter to `Table.same` and my implementation is kind of ugly and bad but I can't think of an obviously better way to do this.

Example usages:

```python
imputed_sex = methods.impute_sex(ds.v.locus(), ds.GT)
```

```python
imputed_sex = methods.impute_sex(ds.v.locus(), 
                                 ds.GT,
                                 aaf=gnomad.AF)
```

### CODE REVIEW

**Comment by danking** [comment]
I actually quite like this with `tablify`. [Take a look at the differences before/after `tablify`](https://github.com/hail-is/hail/pull/2852/commits/31c48bfb8417279b28aa738dab6b56deb0c0b1e2).

File: python/hail/expr/functions.py [hunk 1]
```diff
@@ -515,6 +515,21 @@ def fisher_exact_test(c1, c2, c3, c4):
     return _func("fet", ret_type, c1, c2, c3, c4)
 
 
+@typecheck(x=oneof(expr_float32, expr_float64))
+def floor(x):
```
Comment (tpoterba) [inline python/hail/expr/functions.py]: why don't these accept integer types and do implicit conversion?
```
In [3]: np.floor(2)
Out[3]: 2.0
```

Comment (tpoterba) [reply python/hail/expr/functions.py]: I think there's pretty much nowhere we should accept floats and not ints.

Comment (danking) [reply python/hail/expr/functions.py]: Doesn't it seem like a mistake to call floor on an integer? I can only think of things like floor'ing the return value of an if, but in that case the int will be promoted by the `if` before the `floor` sees it.

Comment (tpoterba) [reply python/hail/expr/functions.py]: I don't think anything in numpy that expects a float would error on int. Shouldn't we do the same?

**Comment by danking** [comment]
@tpoterba back to you, I switched to `expr_numeric` and removed `round` (since that seems controversial now).

We need a final decision about using either:

 - `annotate_all`
 - `tablify`
 - neither, always specify fields as strings

File: src/main/scala/is/hail/table/Table.scala [hunk 2]
```diff
@@ -158,6 +158,26 @@ object Table {
...
+
+  def sameWithinTolerance(t: Type, l: Array[Row], r: Array[Row], tolerance: Double): Boolean = {
+    val used = new Array[Boolean](r.length)
+    var i = 0
+    while (i < l.length) {
+      val li = l(i)
+      var matched = false
+      var j = 0
+      while (!matched && j < l.length && !used(j)) {
+        matched = t.valuesSimilar(li, r(j), tolerance)
```
Comment (jigold) [inline src/main/scala/is/hail/table/Table.scala]: I don't understand why you are doing a product comparison here. Shouldn't it be a zip operation?

File: python/hail/methods/statgen.py [hunk 3]
```diff
@@ -1144,6 +1287,11 @@ def ld_matrix(dataset, force_local=False):
 
     .. include:: ../_templates/req_biallelic.rst
 
+    .. testsetup::
+
+        dataset = vds.annotate_samples_expr('sa = drop(sa, qc)').to_hail2()
```
Comment (jigold) [inline python/hail/methods/statgen.py]: Why is this here?

File: python/hail/expr/functions.py [hunk 4]
```diff
@@ -577,6 +577,16 @@ def fisher_exact_test(c1, c2, c3, c4):
     return _func("fet", ret_type, c1, c2, c3, c4)
 
 
+@typecheck(x=expr_numeric)
+def floor(x):
+    """The largest integral value that is less than or equal to `x`."""
+    return _func("floor", x.dtype, x)
```
Comment (jigold) [inline python/hail/expr/functions.py]: Add doctest examples and Return type.

File: python/hail/__init__.py [hunk 5]
```diff
@@ -41,6 +42,7 @@
            'methods',
            'eval_expr',
            'eval_expr_typed',
+           'tablify',
```
Comment (jigold) [inline python/hail/__init__.py]: This is the wrong module. Also, I don't think this should be in the hail or hail2 package.

File: python/hail/methods/statgen.py [hunk 6]
```diff
@@ -104,6 +104,149 @@ def ibd(dataset, maf=None, bounded=True, min=None, max=None):
...
+    .. include:: ../_templates/req_tvariant.rst
+
+    .. include:: ../_templates/req_biallelic.rst
+
+    Examples
+    --------
+
+    Remove samples where imputed sex does not equal reported sex:
+
+    >>> imputed_sex = methods.impute_sex(ds.rows_table().v.locus(), ds.GT)
```
[the remark this answers is not in the document]
Comment (danking) [reply python/hail/methods/statgen.py]: fixed

File: python/hail/methods/statgen.py [hunk 7]
```diff
@@ -104,6 +104,149 @@ def ibd(dataset, maf=None, bounded=True, min=None, max=None):
...
+    10. A sex is assigned to each sample with the following criteria:
+        - Female when ``F < 0.2``
+        - Male when ``F > 0.8``
+        Use `female-threshold` and `male-threshold` to change this behavior.
+
+    **Annotations**
+
+    The returned column-key indexed :class:`.Table` has the following fields:
+
+    - **s** -- The column key of `call`
```
[the remark this answers is not in the document]
Comment (danking) [reply python/hail/methods/statgen.py]: definitely. to make the code below do this, I just need a way to refer to the column keys. I think tim is gonna add such functionality when he adds multiple column keys

[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/methods/statgen.py]: they exist!

[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/methods/statgen.py]: row keys coming in hot, but col keys exist now. `ds.col_key` is the list of strs

File: python/hail/methods/statgen.py [hunk 8]
```diff
@@ -104,6 +104,149 @@ def ibd(dataset, maf=None, bounded=True, min=None, max=None):
...
+    - **is_female** (:class:`.TBoolean`) -- True if the imputed sex is female, false if male, missing if undetermined
+    - **Fstat** (:class:`.TFloat64`) -- Inbreeding coefficient
+    - **nTotal** (:class:`.TInt64`) -- Total number of variants considered
+    - **nCalled**  (:class:`.TInt64`) -- Number of variants with a genotype call
```
[the remark this answers is not in the document]
Comment (danking) [reply python/hail/methods/statgen.py]: 👍

File: python/hail/methods/statgen.py [hunk 9]
```diff
@@ -104,6 +104,149 @@ def ibd(dataset, maf=None, bounded=True, min=None, max=None):
...
+        - Male when ``F > 0.8``
+        Use `female-threshold` and `male-threshold` to change this behavior.
+
+    **Annotations**
+
+    The returned column-key indexed :class:`.Table` has the following fields:
+
+    - **s** -- The column key of `call`
+    - **is_female** (:class:`.TBoolean`) -- True if the imputed sex is female, false if male, missing if undetermined
+    - **Fstat** (:class:`.TFloat64`) -- Inbreeding coefficient
```
[the remark this answers is not in the document]
Comment (danking) [reply python/hail/methods/statgen.py]: 👍

File: python/hail/methods/statgen.py (hunk 3 above)
Comment (danking) [reply python/hail/methods/statgen.py]: must be a rebase screw up, removed

File: src/main/scala/is/hail/table/Table.scala (hunk 2 above)
Comment (danking) [reply src/main/scala/is/hail/table/Table.scala]: IIUC, the array of rows here are all the values that had the same key in the left table and the right table, respectively.

So I don't know ahead of time the right ordering of the values on the left and the values on the right such that they're similar. For example, if the left had 1.0 and 2.0 and right had 1.9 and 0.9, and the tolerance was 1.0, I'd want to match the 1.0 with the 0.9 and the and 1.9 with the 2.0. But to find that particular set of pairs, I, in general, need to check all possible set of pairs.

**Comment by danking** [comment]
I don't like
```python
ds = ds.annotate_rows(locus2 = crosswalk[ds.rsid].locus)
ds = ds.annotate_rows(AF = gnomad[ds.locus2].AF)
sex = methods.impute_sex(ds.locus2, ds.GT, ds.AF)
```
because it's three lines compared to
```python
sex = methods.impute_sex(crosswalk.locus, ds.GT, aaf=gnomad.AF)
```

That's really the only reason. If everyone thinks this is too complicated, I guess it's fine to go with something else, but it seems like a succinct way to describe the computation.

**Comment by danking** [comment]
I guess if we can think of better documentation, then the current implementation works with either of those expressions.

File: python/hail/methods/statgen.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (danking) [reply python/hail/methods/statgen.py]: ok I think I fixed this to support multiple column keys now

File: python/hail/methods/statgen.py [hunk 10]
```diff
@@ -104,6 +104,149 @@ def ibd(dataset, maf=None, bounded=True, min=None, max=None):
...
+
+    We have used the same implementation as `PLINK v1.7
+    <http://pngu.mgh.harvard.edu/~purcell/plink/summary.shtml#sexcheck>`__.
+
+    Let `gr` be the the genome reference of the type of `locus` (as given by
+    :meth:`.TLocus.reference_genome`)
+
+    1. Filter the dataset to loci on the X contig defined by `gr`.
+
+    4. Calculate alternate allele frequency (AAF) for each row from the dataset.
```
[the remark this answers is not in the document]
Comment (danking) [reply python/hail/methods/statgen.py]: foxed

File: python/hail/expr/functions.py (hunk 4 above)
Comment (danking) [reply python/hail/expr/functions.py]: 👍

File: python/hail/__init__.py (hunk 5 above)
Comment (danking) [reply python/hail/__init__.py]: I don't understand. I figured `tablify` would be generally useful to library writers.

**Comment by danking** [comment]
Is "singly-indexed" expression better than "Table-like expression"?

File: python/hail/expr/functions.py [hunk 11]
```diff
@@ -577,6 +577,40 @@ def fisher_exact_test(c1, c2, c3, c4):
...
+    .. doctest::
+
+        >>> eval_expr(functions.floor(3.1))
+        3.0
+
+    Returns
+    -------
+    :obj:`.Float64Expression`
+    """
+    return _func("floor", x.dtype, x)
```
[the remark this answers is not in the document]
Comment (danking) [reply python/hail/expr/functions.py]: Now I'm back to thinking this really should be `expr_float`, how do I specify the return type correctly without duplicating the promotion logic?

[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/expr/functions.py]: `ret_type = unify_types(x.dtype, TFloat32())`

[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/expr/functions.py]: should give you int32/64 -> float32, float64 ->float64

File: python/hail/methods/statgen.py [hunk 12]
```diff
@@ -104,6 +104,148 @@ def ibd(dataset, maf=None, bounded=True, min=None, max=None):
...
 
 @handle_py4j
+@typecheck(locus=expr_locus,
+           call=expr_call,
+           aaf_threshold=numeric,
+           include_par=bool,
+           female_threshold=numeric,
+           male_threshold=numeric,
+           aaf=nullable(expr_numeric))
+def impute_sex(locus, call, aaf_threshold=0.0, include_par=False, female_threshold=0.2, male_threshold=0.8, aaf=None):
```
[the remark this answers is not in the document]
Comment (danking) [reply python/hail/methods/statgen.py]: This method works perfectly fine on a locus-only dataset, why impose an unnecessary pre-condition?

[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/methods/statgen.py]: we have a whole suite of genetics methods which expect a certain structure for the data. This is how we keep some of our high level of abstraction, while still being totally general. We get a lot more noisy and less user-friendly if we suddenly start requiring the user to pass the locus/alleles field into every method that wants them -- the implementation here is also inconsistent with the other genetics methods like `export_vcf` or `sample_qc`

[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/methods/statgen.py]: if you think we're doing the wrong thing, we should discuss. But everything should certainly be consistent either way

**Comment by cseed** [comment]
I wanted to chime in on this briefly because I think it a good example use case and its design will influence many future methods, so it is important to get the design right.  Thoughts:

 - the underscore stuff is a non-starter in my opinion, and too clever by half.  A lot of my feedback on your stuff is guided by the general heuristic that you should start by writing down the code you want, and then decide how to implement.  You'd never want to write this _ stuff if you didn't have to.

 - I'm still not quite sure what tablify does (in part because the name is too clever by half and in part because it doesn't appear to always return tables).

 - But I think the idea of tablify is something we want, which is to convert (possibly indexed expressions) back into relational objects (Table, MatrixTable) because the latter support a wider set of operations and don't have the "source mismatch problem".  Tim and I discussed this yesterday and we suggest the following interface:

```
  t = build_table()
    .set_globals(x = 5, batch = batch)
    .set_rows(locus = locus, aaf = aaf)
    .build()
```

and

```
  mt = build_table_matix()
    .set_globals(dataset = dataset)
    .set_rows(locus = locus, aaf = aaf)
    .set_entries(GT = GT)
    .build()
```

where the input expressions for each part must all come from the same source (or be compatible, e.g., constants) and the resulting (matrix) table inherits the keys from the original table.  I think there is an unresolved question about how to handle potential name conflicts (e.g. a column key named locus).

**Comment by danking** [comment]
@tpoterba @jigold back to you

File: python/hail/methods/statgen.py [hunk 13]
```diff
@@ -104,6 +103,140 @@ def ibd(dataset, maf=None, bounded=True, min=None, max=None):
...
+      false if male, missing if undetermined.
+    - **f_stat** (:class:`.TFloat64`) -- Inbreeding coefficient.
+    - **n_called**  (:class:`.TInt64`) -- Number of variants with a genotype call.
+    - **expected_homs** (:class:`.TFloat64`) -- Expected number of homozygotes.
+    - **observed_homs** (:class:`.TInt64`) -- Observed number of homozygotes.
+
+    call : :class:`.CallExpression`
+        A genotype call for each row and column. The source dataset's row keys
+        must be [[locus], alleles] with types :class:`.TLocus` and
+        :class:`.ArrayStringExpression`. Moreover, the keys must all be biallelic.
```
Comment (jigold) [inline python/hail/methods/statgen.py]: I think it would be better to say the length of the row key alleles must be equal to 2 (biallelic).

File: python/hail/methods/statgen.py [hunk 14]
```diff
@@ -104,6 +103,140 @@ def ibd(dataset, maf=None, bounded=True, min=None, max=None):
...
+        Samples are called males if F > male_threshold.
+    aaf : :obj:`str` or :obj:`None`
+        A field defining the alternate allele frequency for each row. If
+        ``None``, AAF will be computed from `call`.
+
+    Return
+    ------
+    :class:`.Table`
+        Sex imputation statistics per sample.
+
```
Comment (jigold) [inline python/hail/methods/statgen.py]: delete empty line.

File: python/hail/methods/statgen.py [hunk 15]
```diff
@@ -104,6 +103,140 @@ def ibd(dataset, maf=None, bounded=True, min=None, max=None):
...
+    ds = require_variant(ds, 'impute_sex')
+    def locus(ds):
+        # FIXME use keys
+        return ds.v.locus() # ds[ds.row_key[0]]
+    gr = locus(ds).dtype.reference_genome
+
+    if (aaf is None):
+        ds = ds.annotate_rows(
+            aaf=(agg.sum(ds.call.num_alt_alleles()).to_float64() /
+                 agg.count_where(f.is_defined(ds.call)) / 2))
```
Comment (jigold) [inline python/hail/methods/statgen.py]: should this be `* 2` rather than `/ 2`?

File: python/hail/methods/statgen.py [hunk 16]
```diff
@@ -104,6 +103,140 @@ def ibd(dataset, maf=None, bounded=True, min=None, max=None):
...
+        A field defining the alternate allele frequency for each row. If
+        ``None``, AAF will be computed from `call`.
+
+    Return
+    ------
+    :class:`.Table`
+        Sex imputation statistics per sample.
+
+    """
+
```
Comment (jigold) [inline python/hail/methods/statgen.py]: assert aaf_threshold between [0, 1].

File: python/hail/methods/statgen.py [hunk 17]
```diff
@@ -104,6 +103,140 @@ def ibd(dataset, maf=None, bounded=True, min=None, max=None):
...
+    10. A sex is assigned to each sample with the following criteria:
+        - Female when ``F < 0.2``
+        - Male when ``F > 0.8``
+        Use `female_threshold` and `male_threshold` to change this behavior.
+
+    **Annotations**
+
+    The returned column-key indexed :class:`.Table` has the following fields:
+
+    - **s** -- The column key of `call`.
```
Comment (jigold) [inline python/hail/methods/statgen.py]: I'm not sure how to fix this, but `s` is not always the name of the field. I think we should just delete `**s** --` and it should just be `The column key(s) of call`.

File: python/hail/methods/tests.py [hunk 18]
```diff
@@ -82,6 +82,47 @@ def compare(ds, min=None, max=None):
...
+        vcf_file = utils.get_URI(utils.new_temp_file(prefix="plink", suffix="vcf"))
+        out_file = utils.get_URI(utils.new_temp_file(prefix="plink"))
+
+        methods.export_vcf(ds, vcf_file)
+
+        try:
+            out = sp.check_output(
+                ["plink", "--vcf", vcf_file, "--const-fid", "--check-sex",
+                 "--silent", "--out", out_file],
+                stderr=sp.STDOUT)
```
Comment (jigold) [inline python/hail/methods/tests.py]: Can you add a test using the alternate allele frequency expression computed from call_stats.AF and make sure your results agree with your formula with aaf=None?

File: python/hail/methods/statgen.py (hunk 15 above)
Comment (danking) [reply python/hail/methods/statgen.py]: divison is left associative so no, but I rewrote it to
```
(... /
 (2.0 * ...))
```
so hopefully that's more clear

Comment (tpoterba) [reply python/hail/methods/statgen.py]: can we use call stats?

File: python/hail/methods/statgen.py [hunk 19]
```diff
@@ -179,7 +179,8 @@ def impute_sex(call, aaf_threshold=0.0, include_par=False, female_threshold=0.2,
     call : :class:`.CallExpression`
         A genotype call for each row and column. The source dataset's row keys
         must be [[locus], alleles] with types :class:`.TLocus` and
-        :class:`.ArrayStringExpression`. Moreover, the keys must all be biallelic.
+        :class:`.ArrayStringExpression`. Moreover, the alleles array have
```
Comment (jigold) [inline python/hail/methods/statgen.py]: alleles array must have exactly two elements (biallelic).

File: python/hail/methods/statgen.py [hunk 20]
```diff
@@ -212,9 +213,7 @@ def locus(ds):
     gr = locus(ds).dtype.reference_genome
 
     if (aaf is None):
-        ds = ds.annotate_rows(
-            aaf=(agg.sum(ds.call.num_alt_alleles()).to_float64() /
-                 agg.count_where(f.is_defined(ds.call)) / 2))
+        ds = ds.annotate_rows(aaf=agg.call_stats(ds.call, ds.v).AF[0])
```
Comment (jigold) [inline python/hail/methods/statgen.py]: I think this should be `AF[1]`.

File: python/hail/methods/tests.py [hunk 21]
```diff
@@ -123,6 +123,10 @@ def test_impute_sex_same_as_plink(self):
 
         self.assertTrue(plink_sex._same(sex, tolerance=1e-3))
 
+        ds = ds.annotate_rows(aaf=(agg.call_stats(ds.GT, ds.v)).AF[0])
```
Comment (jigold) [inline python/hail/methods/tests.py]: same -- AF[1]. I double checked and the answer won't change if you use AF[0] or AF[1] because it's using `p(1-p)`. However, I'd prefer to be consistent if the parameter name is alternate allele frequency.

File: python/hail/methods/statgen.py [hunk 22]
```diff
@@ -196,8 +197,8 @@ def impute_sex(call, aaf_threshold=0.0, include_par=False, female_threshold=0.2,
     ------
     :class:`.Table`
         Sex imputation statistics per sample.
-
     """
+    assert(aaf_threshold >= 0.0 and aaf_threshold <= 1.0)
```
Comment (jigold) [inline python/hail/methods/statgen.py]: Sorry, should have been clearer what I wanted. Something like this. Not sure what error to raise.
```
if not 0.0 <= aaf_threshold <= 1.0:
    raise FatalError("invalid argument for `aaf_threshold`. Must be in range [0, 1].")
```

**Comment by tpoterba** [comment]
rebase conflict after merging splitmulti
